plugin-grace-hopper/src/lib.rs

```
@@ -0,0 +1,404 @@
+mod probe;
+
+use anyhow::Context;
+use probe::GraceHopperProbe;
+use serde::{Deserialize, Serialize};
+use std::{
+    fs::{self, File},
+    io::{self, BufRead},
+    path::PathBuf,
+    time::Duration,
+};
+
+use alumet::{
+    pipeline::elements::source::trigger::TriggerSpec,
+    plugin::{
+        rust::{deserialize_config, serialize_config, AlumetPlugin},
+        ConfigTable,
+    },
+};
+
+pub struct GraceHopperPlugin {
+    config: Config,
+}
+
+impl AlumetPlugin for GraceHopperPlugin {
+    fn name() -> &'static str {
+        "grace-hopper"
+    }
+
+    fn version() -> &'static str {
+        env!("CARGO_PKG_VERSION")
+    }
+
+    fn init(config: ConfigTable) -> anyhow::Result<Box<Self>> {
+        let config = deserialize_config(config)?;
+        Ok(Box::new(GraceHopperPlugin { config }))
+    }
+
+    fn default_config() -> anyhow::Result<Option<ConfigTable>> {
+        let config = serialize_config(Config::default())?;
+        Ok(Some(config))
+    }
+
+    fn start(&mut self, alumet: &mut alumet::plugin::AlumetPluginStart) -> anyhow::Result<()> {
+        let base_dir = self.config.root_path.to_string();
+        // Try to open the directory
+        if let Ok(entries) = fs::read_dir(base_dir) {
+            for entry in entries {
+                if let Ok(entry) = entry {
+                    let path = entry.path();
+                    // Check if it's a directory
+                    if path.is_dir() {
```

I think that a fail fast would be more readable